### PR TITLE
Add ability to run per key inference

### PR DESCRIPTION
**Pull request by damccorm**
This builds on top of #27603 to add support for running inference against a cohort of different models using the KeyedModelHandler. Full design is here - https://docs.google.com/document/d/1kj3FyWRbJu1KhViX07Z0Gk0MU0842jhYRhI-DMhhcv4/edit#heading=h.wskna8eurvjv

This piece adds basic support. In future prs we will need to cover:

- Example + an E2E test (I have run this in an E2E pipeline, but decided to split the example out from this pr for size reasons)
- Limiting the maximum number of models that can be held in memory
- Model updates support
- Metrics improvements
- Docs

Given these gaps, I don't think we should merge before the release cut on Wednesday, but we can get this in after and iterate with small PRs

Part of #27628

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by AnandInguva** [comment]
assign to next reviewer

**Comment by damccorm** [comment]
@tvalentyn FYI, here's an integration test built on top of this PR - https://github.com/apache/beam/pull/27880

I'm keeping it separate for now to keep this one easier to review

File: sdks/python/apache_beam/ml/inference/base.py [hunk 1]
```diff
@@ -243,70 +244,273 @@ def share_model_across_processes(self) -> bool:
...
+      for inf in mh.run_inference(unkeyed_batches, keyed_model, inference_args):
+        predictions.append((key, inf))
+      keyed_model_shared_handle.release(keyed_model)
+
+    return predictions
 
   def get_num_bytes(self, batch: Sequence[Tuple[KeyT, ExampleT]]) -> int:
-    keys, unkeyed_batch = zip(*batch)
-    return len(pickle.dumps(keys)) + self._unkeyed.get_num_bytes(unkeyed_batch)
+    if not self._many_models:
```
Comment (tvalentyn) [inline sdks/python/apache_beam/ml/inference/base.py:468]: given how frequently this appears, consider `if self._single_model` instead

File: sdks/python/apache_beam/ml/inference/base.py [hunk 2]
```diff
@@ -243,70 +244,273 @@ def share_model_across_processes(self) -> bool:
...
+      keys, unkeyed_batch = zip(*batch)
+      return len(
+          pickle.dumps(keys)) + self._unkeyed.get_num_bytes(unkeyed_batch)
+    return len(pickle.dumps(batch))
```
Comment (tvalentyn) [inline sdks/python/apache_beam/ml/inference/base.py:472]: i wonder why this wouldn't work for all modelhandlers. but as is, we'd have an (undocumented?) difference in behavior that we don't delegate estimation to corresponding MHs.

File: sdks/python/apache_beam/ml/inference/base.py [hunk 3]
```diff
@@ -243,70 +244,273 @@ def share_model_across_processes(self) -> bool:
...
+    for key, unkeyed_batches in batch_by_key.items():
+      id = self._key_to_id_map[key]
+      mh = self._id_to_mh_map[id]
+      keyed_model_tag = model.load(id)
+      keyed_model_shared_handle = multi_process_shared.MultiProcessShared(
+          mh.load_model, tag=keyed_model_tag)
+      keyed_model = keyed_model_shared_handle.acquire()
+      for inf in mh.run_inference(unkeyed_batches, keyed_model, inference_args):
+        predictions.append((key, inf))
+      keyed_model_shared_handle.release(keyed_model)
```
[the remark this answers is not in the document]
Comment (tvalentyn) [reply sdks/python/apache_beam/ml/inference/base.py:463]: Do we plan to adjust batch generation logic to reorganize the keys?

File: sdks/python/apache_beam/ml/inference/base.py [hunk 4]
```diff
@@ -243,70 +244,273 @@ def share_model_across_processes(self) -> bool:
...
+      return self._unkeyed.update_model_path(model_path=model_path)
+    if model_path is not None:
+      raise RuntimeError(
+          'Model updates are currently not supported for ' +
+          'KeyedModelHandlers with multiple different per-key ' +
+          'ModelHandlers.')
 
   def get_preprocess_fns(self) -> Iterable[Callable[[Any], Any]]:
-    return self._unkeyed.get_preprocess_fns()
+    return []
```
Comment (tvalentyn) [inline sdks/python/apache_beam/ml/inference/base.py:505]: what's the rationale for this change?

File: sdks/python/apache_beam/ml/inference/base.py [hunk 5]
```diff
@@ -243,70 +244,273 @@ def share_model_across_processes(self) -> bool:
     return False
 
 
+class _ModelManager:
```
[the remark this answers is not in the document]
Comment (damccorm) [reply sdks/python/apache_beam/ml/inference/base.py:247]: That's right

File: sdks/python/apache_beam/ml/inference/base.py (hunk 2 above)
Comment (damccorm) [reply sdks/python/apache_beam/ml/inference/base.py:484]: I think some model handlers can provide tight bounds on batch sizes without doing a full pickling of them. I added shelling out to model handlers as an improvement in #27628 since I think we can may be able to do that per key more efficiently. Or we can just use any of the model handlers since they're all taking the same input type anyways

File: sdks/python/apache_beam/ml/inference/base.py (hunk 4 above)
Comment (damccorm) [reply sdks/python/apache_beam/ml/inference/base.py:517]: We don't allow these to have a non-zero length anyways (see the constructor). Them being non-empty was pointless before anyways, but I wanted to avoid having to special case the different types of unkeyed here.

File: sdks/python/apache_beam/ml/inference/base.py [hunk 6]
```diff
@@ -243,70 +244,273 @@ def share_model_across_processes(self) -> bool:
...
+      raise ValueError(
+          "Cannot increment max_models if self._max_models is None (unlimited" +
+          " models mode).")
+    self._max_models += increment
+
+
+# Use a dataclass instead of named tuple because NamedTuples and generics don't
+# mix well across the board for all versions:
+# https://github.com/python/typing/issues/653
+class KeyMhMapping(Generic[KeyT, ExampleT, PredictionT, ModelT]):
```
[the remark this answers is not in the document]
Comment (damccorm) [reply sdks/python/apache_beam/ml/inference/base.py:323]: I added a docstring, I don't think anything that doesn't have a ModelHandler reference is right though since they're mapping keys to model handlers

File: sdks/python/apache_beam/ml/inference/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (damccorm) [reply sdks/python/apache_beam/ml/inference/base.py:475]: Good idea, done!

File: sdks/python/apache_beam/ml/inference/base.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tvalentyn) [reply sdks/python/apache_beam/ml/inference/base.py:323]: I agree it's less precise; but ultimately users care about the key->model mapping, and MH is an intermediary. I was trying to come up with a more readable and speakable name , esp since it's public api; also it's more common in python to cap all letters in the acronym (e.g. HTTP vs Http, but i i'd rather have MhM than MHM in this case i think...). Sadly i don't have a good answer. Maybe others have ideas.

File: sdks/python/apache_beam/ml/inference/base.py [hunk 7]
```diff
@@ -243,70 +244,273 @@ def share_model_across_processes(self) -> bool:
...
+    model handlers corresponding to the keys for which that ModelHandler should
+    be used. For example, the following configuration could be used to map keys
+    1-3 to ModelHandler1 and keys 4-5 to ModelHandler2:
+
+        k1 = ['k1', 'k2', 'k3']
+        k2 = ['k4', 'k5']
+        KeyedModelHandler([KeyMhMapping(k1, mh1), KeyMhMapping(k2, mh2)])
+
+    Note that a single copy of each of these models may all be held in memory
+    at the same time; be careful not to load too many large models or your
```
[the remark this answers is not in the document]
Comment (tvalentyn) [reply sdks/python/apache_beam/ml/inference/base.py:359]: consider adding a TODO for reference in case someone comes to read the docstring before external reference is added.

File: sdks/python/apache_beam/ml/inference/base.py (hunk 4 above)
Comment (tvalentyn) [reply sdks/python/apache_beam/ml/inference/base.py:517]: will this have an impact on chain calling into unkeyed preprocess funs for single or multi-model case?

Comment (tvalentyn) [reply sdks/python/apache_beam/ml/inference/base.py:517]: note that this is also a default implementation in the superclass.

**Review by tvalentyn — APPROVED** [review]

File: sdks/python/apache_beam/ml/inference/base.py (hunk 4 above)
Comment (damccorm) [reply sdks/python/apache_beam/ml/inference/base.py]: > will this have an impact on chain calling into unkeyed preprocess funs for single or multi-model case?

Yeah, though not significantly. It doesn't allow you to strictly chain unkeyed functions if you have a KeyedModelHandler, though you can pretty easily do things like: `with_preprocess_fn(lambda k, v: k, my_func(v))`. Note that this PR doesn't change that behavior

> note that this is also a default implementation in the superclass, so we can remove it entirely

Good call, done

File: sdks/python/apache_beam/ml/inference/base.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (damccorm) [reply sdks/python/apache_beam/ml/inference/base.py:323]: That's what I ended up with in the design doc so I think I'll stick with it for now. If we come up with something better we do have 6 weeks before the next cut

File: sdks/python/apache_beam/ml/inference/base.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (damccorm) [reply sdks/python/apache_beam/ml/inference/base.py:359]: Done
